Basic support for editable missions

jeredw opened the pull request:
This is an early effort at editable missions - functional, but not polished.

I'm sending it out now because the most recent meeting notes mention data model questions, and this branch involves some important migrations. Those seem to be expensive and serial in our workflow, so I'd like to forestall conflicts.

Users can now copy missions, edit their titles/descriptions and rearrange goals. When starting a mission, we clone it and mark it "frozen", which will also simplify reconstructing history later.

Still todo:
- ACLs for ownership.
- Filtering missions to a reasonable set on the team page.
- Searching for and adding goals (including the notion of "goal categories").

jeredw committed:
migrate to missiongoal join table

jeredw committed:
migrate to missiongoal join table

jeredw committed:
clone/freeze missions once started

this will simplify editing and history

jeredw committed:
clone/freeze missions once started

this will simplify editing and history

jeredw committed:
use mission.started_at instead of team field

jeredw committed:
use mission.started_at instead of team field

jeredw committed:
peewee-moves add_column uses str for type

jeredw committed:
peewee-moves add_column uses str for type

jeredw committed:
editable mission name and description

jeredw committed:
editable mission name and description

jeredw committed:
add confirm modal to start button

jeredw committed:
add confirm modal to start button

jeredw committed:
basic ui for goal rescheduling

jeredw committed:
basic ui for goal rescheduling

jeredw committed:
allow editing mission duration

jeredw committed:
allow editing mission duration

jeredw committed:
swap primary and secondary for clearer sidedness

jeredw committed:
swap primary and secondary for clearer sidedness

jeredw committed:
less clashy toolbar background?

jeredw committed:
less clashy toolbar background?

jeredw committed:
add goal categories to facilitate searching

jeredw committed:
add goal categories to facilitate searching

jeredw committed:
placeholder goal search function

jeredw committed:
placeholder goal search function

jeredw committed:
remove dangling empty weeks

jeredw committed:
remove dangling empty weeks

jeredw committed:
move goals to first open week

jeredw committed:
move goals to first open week

jeredw committed:
only clear dangling weeks for move up

jeredw committed:
only clear dangling weeks for move up

jeredw committed:
more muted color for category label

jeredw committed:
more muted color for category label

igor47 commented on migrations/0002_add_name_to_user.py:
```diff
@@ -2,10 +2,8 @@
-
 def upgrade(migrator):
-    migrator.add_column('user', 'name', CharField, max_length=255, default='')
+    migrator.add_column('user', 'name', 'char', max_length=255, default='')
```
did you mean to edit this old migration?

jeredw replied:
Yeah this fix made it in with this branch since I noticed when testing my new migrations. (This param is supposed to be a str looked up in a dictionary of str -> type, where CharField is the default if no string matches.)

igor47 commented on migrations/0016_drop_unused_mission_and_goal_fields.py:
```diff
@@ -0,0 +1,26 @@
+"""
+
+def add_foreign_key(migrator, from_table, from_field, to_table, to_field, **kwargs):
+  """Add a foreign key, which playhouse-moves doesn't seem to do properly."""
```
huh that's funny. how did it work in the last migration that added the key constraints?

jeredw replied:
I used the same helper function last time, after reading the peewee-moves code and discovering this didn't work. I tried for 30 minutes to reuse the same function somehow, including an insane attempt to make relative imports work for things beginning with numbers, but thought better of it and gave up. peewee-moves is not my favorite library.

igor47 commented on spaceship/templates/mission.html:
```diff
@@ -26,75 +77,284 @@ <h3>{{ mission.title }}</h3>
+      <button type="button" class="remove-goal btn btn-primary" title="Remove"><i class="fas fa-times-circle"></i></button>
+    </div>
+  </div>`);
+  $(dom).find('.remove-goal').click(function() {
```
we should really start using something like `vuejs` to make a lot of this dom manipulation declarative

jeredw replied:
Hrm looks interesting, I will look into it. It is seeming like we are going to have a fair bit of js here after all and jquery is feeling awfully crusty...

igor47 commented on spaceship/views.py:
```diff
@@ -706,6 +750,26 @@ def edit():
+      mission = Mission.get(Mission.id == object_id)
+    except Mission.DoesNotExist:
+      return jsonify({'ok': False})
+    # TODO acl
```
lol. watch out, i'm coming to edit your mission!

jeredw replied:
nooo

igor47 reviewed (approved):
the backend looks very sensible. how we do the UI is still up in the air from a product perspective -- at the last meeting, we talked about minimum viable customizability to avoid confusing people. like, maybe a slider you drag for "difficulty level" and then it selects your goals for you, on easy you would get "eat less beef" and on hard you might get "go totally vegan"  -- but either way we need the backend infrastructure for customizing missions.

from a javascript perspective, i think some declarative js components would help. i've been a vue fanboy because i like that you can add it progressively without going all-in the way you have to with react, curious what others think.

jeredw committed:
stray semicolon

jeredw committed:
stray semicolon

jeredw committed:
add ability to set host for run.flask

(i need this because my new cheap chromebook (very slowly) runs linux in
a vm, where i want to run the server on 0.0.0.0 so i can get at it from
the host)

jeredw committed:
Merge branch 'mission-builder' of github.com:spaceshipearth/pyspaceship into mission-builder
